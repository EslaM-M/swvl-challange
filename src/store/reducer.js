import {
  AddBooking,
  UpdateRouteDistance,
  UpdateLocation,
  StartTrip,
  UpdateStation,
  UpdateFullDistance,
  LoadState,
  EndTrip,
  RestTrip,
  UpdateTripInfo
} from "./Actions";
export const initialState = {
  bookings: [],
  bookingAllowed: true,
  route: [],
  captainLocation: {
    lat: null,
    lng: null
  },
  tripStatus: "NOT_STARTED",
  tripStartTime: null,
  arrivalStatus: "",
  currentStationId: null,
  fullDistance: null,
  baseFare: {
    price: "",
    currency: ""
  },
  startPoint: "",
  endPoint: "",
  driver: {
    name: "",
    rating: "",
    bus: {
      modal: "",
      plate: ""
    }
  }
};


export function reducer(state, action) {
  switch (action.type) {
    case "ADD_BOOKING":
      return AddBooking(state, action.payload);
    case "UPDATE_ROUTE_DISTANCE":
      return UpdateRouteDistance(state, action.payload);
    case "UPDATE_LOCATION":
      return UpdateLocation(state, action.payload);
    case "START_TRIP":
      return StartTrip(state, action.payload);
    case "UPDATE_TRIP_INFO":
      return UpdateTripInfo(state, action.payload);
    case "UPDATE_STATION":
      return UpdateStation(state, action.payload);
    case "UPDATE_FULL_DISTANCE":
      return UpdateFullDistance(state, action.payload);
    case "END_TRIP":
      return EndTrip(state, action.payload);
    case "LOAD_STATE":
      return LoadState(state,action.payload);
    case "REST_TRIP":
      return RestTrip(state);
    default:
      return state;
  }
}
