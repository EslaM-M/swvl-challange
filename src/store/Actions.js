import Users from "../assets/data/users";
import TripInformation from "../assets/data/trip-information";
import BookingStatus from "../assets/data/booking-status";
import { initialState } from "./reducer";

export const AddBooking = (state, payload) => {
  if (state.bookings.length === 12) {
    return state;
  }
  payload = {
    ...payload,
    userName: Users[state.bookings.length].name,
    userImage: Users[state.bookings.length].image,
    userId: Math.random()
      .toString()
      .substr(2, 4),
    status: "Booked"
  };
  let bookingAllowed = true;
  if (state.bookings.length === 11) {
    bookingAllowed = false;
  }

  const stationsUpdate = state.route.map(e => {
    if (parseInt(e.stationId) === parseInt(payload.pickupStation)) {
      e.count++;
    }
    return e;
  });
  return {
    ...state,
    bookings: [...state.bookings, payload],
    bookingAllowed: bookingAllowed,
    route: stationsUpdate
  };
};

export const UpdateRouteDistance = (state, payload) => {
  return {
    ...state,
    route: [...payload],
    captainLocation: {
      ...state.captainLocation,
      lat: payload[0].lat,
      lng: payload[0].lng
    }
  };
};

export const UpdateLocation = (state, payload) => {
  const newState = {
    ...state,
    captainLocation: {
      ...state.captainLocation,
      lat: payload.lat,
      lng: payload.lng
    }
  };
  localStorage.setItem("root", JSON.stringify(newState));
  return newState;
};

export const LoadState = (state, payload) => {
  let newState = JSON.parse(localStorage.getItem("root"));
  if (!newState) {
    newState = {
      ...initialState,
      route: payload.route.map(e => {
        e.count = 0;
        return e;
      }),
      startPoint: payload.startPoint,
      tripStatus: "NOT_STARTED",
      endPoint: payload.endPoint,
      driver: payload.driver,
      baseFare: payload.baseFare
    };
  }
  return newState;
};

export const StartTrip = (state, payload) => {
  return {
    ...state,
    tripStatus: "IN_PROGRESS",
    arrivalStatus: payload.arrivalStatus,
    tripStartTime: new Date()
  };
};
export const EndTrip = (state, payload) => {
  let newState = {
    ...state,
    tripStatus: "FINISHED",
    bookings: state.bookings.map(e => {
      if (e.status === "Checked In") {
        e.status = "Compeleted";
      }
      return e;
    })
  };
  localStorage.setItem("root", JSON.stringify(newState));
  return newState;
};

export const RestTrip = (state, payload) => {
  const newState = {
    ...state,
    bookings: [],
    tripStatus: "NOT_STARTED",
    bookingAllowed: true,
    tripStartTime: null,
    arrivalStatus: "",
    captainLocation: {
      lat: TripInformation.route[0].lat,
      lng: TripInformation.route[0].lng
    },
    route: state.route.map(e => {
      e.count = 0;
      return e;
    }),
    currentStationId: null
  };
  localStorage.setItem("root", JSON.stringify(newState));

  return newState;
};

export const UpdateStation = (state, payload) => {
  const newBookings = state.bookings.map(e => {
    if (parseInt(e.pickupStation) === parseInt(payload.id)) {
      e.status = BookingStatus[Math.floor(Math.random() * 3) + 1];
    }
    return e;
  });
  const newRoute = state.route.map(e => {
    if (e.stationId === payload.id) {
      e.arrivalTime = payload.time;
      if (state.arrivalStatus === "early") {
        e.arrivalStatus = ["early", "ontime"][Math.floor(Math.random() * 2)];
      } else if (state.arrivalStatus === "ontime") {
        e.arrivalStatus = ["early", "ontime", "late"][
          Math.floor(Math.random() * 3)
        ];
      } else if (state.arrivalStatus === "late") {
        e.arrivalStatus = ["ontime", "late"][Math.floor(Math.random() * 2)];
      }
    }
    return e;
  });
  return {
    ...state,
    currentStationId: payload.id,
    bookings: [...newBookings],
    route: [...newRoute]
  };
};

export const UpdateFullDistance = (state, payload) => {
  return {
    ...state,
    fullDistance: payload
  };
};
