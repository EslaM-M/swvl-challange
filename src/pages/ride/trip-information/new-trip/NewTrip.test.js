import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewTrip from "./index";
import { initialState } from "../../../../store/reducer";
describe("NewTrip Component", () => {
  let realUseContext;
  let useContextMock;
  // Setup mock
  beforeEach(() => {
    realUseContext = React.useContext;
    useContextMock = React.useContext = jest.fn();
    useContextMock.mockReturnValue({
      state: initialState
    });
  });
  // Cleanup mock
  afterEach(() => {
    React.useContext = realUseContext;
  });

  test("Component snapshot render correctly", () => {
    const container = render(<NewTrip show={true} />);
    expect(container).toMatchSnapshot();
  });

  test("Component save changes disabled if no user input", async () => {
    const { getByText } = render(<NewTrip show={true} />);
    const action = getByText("Save Changes");
    expect(action.disabled).toBe(true);
  });
  test("Component select station have the exact number of options", async () => {
    const { baseElement } = render(<NewTrip show={true} />);
    const options = baseElement.querySelectorAll("option");
    expect(options.length).toBe(4);
  });

  test("Component Save Changes Enabled after add both (station and payment)", async () => {
    const { getByText, getByTestId } = render(<NewTrip show={true} />);
    fireEvent.change(getByTestId("select-status"), {
      target: { value: "early" }
    });
    const action = getByText("Save Changes");
    expect(action.disabled).toBe(false);
  });

  test("Component Save Changes Disabled after payment type ", async () => {
    const { getByText } = render(<NewTrip show={true} />);
    const action = getByText("Save Changes");
    expect(action.disabled).toBe(true);
  });
});
