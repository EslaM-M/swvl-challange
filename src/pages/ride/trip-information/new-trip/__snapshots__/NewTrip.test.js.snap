// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTrip Component Component snapshot render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    style="padding-right: 0px;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="fade modal-backdrop show"
    />
    <div
      aria-hidden="true"
      aria-modal="true"
      class="fade modal show"
      role="dialog"
      style="display: block;"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
        role="document"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <div
              class="modal-title h4"
            >
              New Trip
            </div>
            <button
              class="close"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
              <span
                class="sr-only"
              >
                Close
              </span>
            </button>
          </div>
          <div
            class="modal-body"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                  for="newBookingForm.arrivalStatus"
                >
                  Arrival Status
                </label>
                <select
                  class="form-control"
                  data-testid="select-status"
                  id="newBookingForm.arrivalStatus"
                  name="arrivalStatus"
                >
                  <option
                    value=""
                  >
                    Select Status
                  </option>
                  <option
                    value="early"
                  >
                    Early
                  </option>
                  <option
                    value="ontime"
                  >
                    On Time
                  </option>
                  <option
                    value="late"
                  >
                    Late
                  </option>
                </select>
              </div>
            </form>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary"
              type="button"
            >
              Close
            </button>
            <button
              class="btn btn-primary"
              disabled=""
              type="button"
            >
              Save Changes
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
